### PR TITLE
Not overwrite the whole SystemResourceInfo config map

### PR DESCRIPTION
**Pull request by mqliang**
https://github.com/apache/pinot/pull/8785 put InstanceResourceInfo updating logic into a private method `private void startupServiceStatusCheck(long endTimeMs)` method, which make overriding the InstanceResourceInfo logic impossible.

This PR extracts the logic back into a protect method.

@jackjlli

### CODE REVIEW

File: pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java [hunk 1]
```diff
@@ -751,6 +751,21 @@ public ServerInstance getServerInstance() {
...
+   * @param helixAdmin Helix Admin
+   * @param helixClusterName Name of Helix cluster
+   * @param instanceId Id of instance for which to set the system resource info
+   * @param systemResourceMap Map containing system resource info
+   */
+  protected void setInstanceResourceInfo(HelixAdmin helixAdmin, String helixClusterName, String instanceId,
+      Map<String, String> systemResourceMap) {
+    InstanceConfig instanceConfig = helixAdmin.getInstanceConfig(helixClusterName, instanceId);
+    instanceConfig.getRecord().setMapField(Helix.Instance.SYSTEM_RESOURCE_INFO_KEY, systemResourceMap);
+    helixAdmin.setInstanceConfig(helixClusterName, instanceId, instanceConfig);
```
Comment (jackjlli) [inline pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java:766]: You don't need to call this setter method one more time as it'll be done at the end. Please refer to the other logic in the same class.

Comment (mqliang) [reply pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java]: done

File: pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java [hunk 2]
```diff
@@ -346,7 +346,7 @@ private void updateInstanceConfigIfNeeded(ServerConf serverConf) {
     Map<String, String> systemResourceInfoMap = new SystemResourceInfo().toMap();
     if (!systemResourceInfoMap.equals(znRecord.getMapField(Instance.SYSTEM_RESOURCE_INFO_KEY))) {
```
[the remark this answers is not in the document]
Comment (mqliang) [reply pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java:347]: > It'd be good to check whether the new kv pairs from the new map equal to the ones appeared from ZNRecord.

we already do the check:
```
if (!systemResourceInfoMap.equals(znRecord.getMapField(Instance.SYSTEM_RESOURCE_INFO_KEY)))
```

File: pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java [hunk 3]
```diff
@@ -343,10 +343,17 @@ private void updateInstanceConfigIfNeeded(ServerConf serverConf) {
...
-      LOGGER.info("Updating instance: {} with system resource info: {}", _instanceId, systemResourceInfoMap);
-      znRecord.setMapField(Instance.SYSTEM_RESOURCE_INFO_KEY, systemResourceInfoMap);
+    Map<String, String> newSystemResourceInfoMap = new SystemResourceInfo().toMap();
+    Map<String, String> oldSystemResourceInfoMap =
+        znRecord.getMapField(CommonConstants.Helix.Instance.SYSTEM_RESOURCE_INFO_KEY);
+    if (!newSystemResourceInfoMap.equals(oldSystemResourceInfoMap)) {
+      LOGGER.info("Updating instance: {} with new system resource info: {}", _instanceId, newSystemResourceInfoMap);
+      // oldSystemResourceInfoMap may contains more KV pairs than newSystemResourceInfoMap, we need add the extra KV
+      // pairs into newSystemResourceInfoMap.
+      for (Map.Entry<String, String> entry : oldSystemResourceInfoMap.entrySet()) {
```
Comment (jackjlli) [inline pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java:353]: This logic seems incorrect. You now basically replace the new KV pair with old value.

File: pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java [hunk 4]
```diff
@@ -343,10 +343,21 @@ private void updateInstanceConfigIfNeeded(ServerConf serverConf) {
     }
 
     // Update system resource info (CPU, memory, etc)
-    Map<String, String> systemResourceInfoMap = new SystemResourceInfo().toMap();
-    if (!systemResourceInfoMap.equals(znRecord.getMapField(Instance.SYSTEM_RESOURCE_INFO_KEY))) {
-      LOGGER.info("Updating instance: {} with system resource info: {}", _instanceId, systemResourceInfoMap);
-      znRecord.setMapField(Instance.SYSTEM_RESOURCE_INFO_KEY, systemResourceInfoMap);
+    Map<String, String> newSystemResourceInfoMap = new SystemResourceInfo().toMap();
+    Map<String, String> oldSystemResourceInfoMap =
```
Comment (jackjlli) [inline pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java:347]: sed `s/oldSystemResourceInfoMap/existingSystemResourceInfoMap/`

**Review by jackjlli — APPROVED** [review]

**Pull request merged**